### DISCUSSION TITLE
Extension Request?: Duplicate and paste line

### DISCUSSION THREAD

**Discussion by j-a-s-o-n-g** [microsoft/vscode-discussions#831] [Extension Development QnA] [answered]
I've been looking for an extension that can do this behavior for about 6 months now, and have finally resorted to asking a community if anyone can easily make one. I don't think it's terribly complex in behavior, but I know nothing about making extensions for vscode.

TL:DR: Create a vscode extension that will take the current line and current caret position, and duplicate that line for x amount of carets (or lines) in the clipboard and also insert the contents of the clipboard.

For example, I have this in my clipboard:

```
a
b
c
d
e
```

Say I have this block of code: (where | is the caret)

```
console.log("This is: | ");
```

If I execute the command "Duplicate and Paste" it will generate this:

```
console.log("This is: a ");
console.log("This is: b ");
console.log("This is: c ");
console.log("This is: d ");
console.log("This is: e ");
```

Some edge cases might be text is selected: (Where [] are selection bounds)

```
var omg[This] = true;
```

Execute command, resulting in:

```
var omga = true;
var omgb = true;
var omgc = true;
var omgd = true;
var omge = true;
```

**Comment by alefragnani** [comment]
Hi @j-a-s-o-n-g ,

I think the native [multi-cursor](https://code.visualstudio.com/docs/editor/codebasics#_multiple-selections-multicursor) support  in VS Code allows you to achieve similar productivity. 

1. Paste the same initial content from the clipboard
```
a
b
c
d
e
```

2. Move the cursor prior to **a**
3. Call `Add Cursor Below` command 4x, making all 5 files to have a cursor prior to the letter
4. Start typing `console.log("This is `
5. Hit <kbd>End</kbd>
6. Type ` ");`
7. There you go, the same content

```
console.log("This is: a ");
console.log("This is: b ");
console.log("This is: c ");
console.log("This is: d ");
console.log("This is: e ");
```

Alternatively, you could copy ` console.log("This is: ` to the clipboard before step 2 and paste it in step 4. It would paste the same content on each of the 5 lines.

Otherwise, if **multi-cursor** is not enough for you, you could create your own extension, for sure. Start looking at https://code.visualstudio.com/api/get-started/your-first-extension, with the initial steps to have an extension up and running. Then, you should look the [TextEditor API](https://code.visualstudio.com/api/references/vscode-api#TextEditor) to deal with text manipulation and selection, and probably the [Clipboard API](https://code.visualstudio.com/api/references/vscode-api#Clipboard) as well. Also, the vscode-extensions-sample repo would be a good source of samples as well (https://github.com/microsoft/vscode-extension-samples/tree/main/document-editing-sample)

Hope this helps

**Comment by j-a-s-o-n-g** [comment]
You most definitely can produce this with the extra steps you mentioned. However, it would be great if the extension took out all of the extra steps and mental calculations you need to do in order to achieve such a simple task.

Although you put a lot of effort into your explanation, it does not solve the topic and therefor I cannot mark it as the answer.

**Reply by alefragnani** [reply, marked as answer]
Hi @j-a-s-o-n-g ,

Well, if the native multi-cursor is not suitable for you needs, you can create a new extension for such task, using the documentations and samples I provided above. Of course, if you are willing to develop such extension yourself. Otherwise, you could open a feature request in VS Code repo (https://github.com/microsoft/vscode/issues/new?assignees=&labels=&projects=&template=feature_request.md), and maybe the VS Code team could add this in future releases.

Another alternative is to ask to some extension author to develop such extension. Maybe someone in the [VS Code Extension Slack Channel](https://aka.ms/vscode-dev-community) would be available.

By the way, for future feature requests about VS Code, please use the VS Code repository instead (https://github.com/microsoft/vscode/issues/new/choose). This Discussions repo is focused for extension development for now.

Hope this helps

**Reply by j-a-s-o-n-g** [reply]
This invite link is not valid.

**Reply by alefragnani** [reply]
Sorry, I didn't know it was outdated. This is the channel https://vscode-dev-community.slack.com.

If you can't access it, try this invite (https://vscode-dev-community.slack.com/join/shared_invite/zt-24clr568w-wCxESYPl6s9Q_D7JmwYonw#/shared-invite/email) that I just created. It should be valid for up to 30 days.
